cia.mk

```
@@ -1,59 +1,91 @@
-	ctrtool --seeddb=$(vc_dir)/seeddb.bin \
-	        --romfsdir=$@/manual \
-	        $@/contents.0001.*
-	rm -f $@/contents.*
-	rm -f $@/romfs/rom/*
-	rm -f $@/romfs/*.patch
-
-$(join $(addsuffix /romfs/, $(vc_rom_dir)), $(vc_patch)): $(vc_rom_dir) ./$(vc_patch)
-	cp ./$(@F) $@
-	
-$(join $(addsuffix /romfs/rom/, $(vc_rom_dir)), $(vc_gbc)): $(vc_rom_dir) ./$(vc_gbc)
-	cp ./$(@F) $(basename $@)
-
-$(vc_game_cxi): $(vc_rom_dir) $(join $(addsuffix /romfs/, $(vc_rom_dir)), $(vc_patch)) $(join $(addsuffix /romfs/rom/, $(vc_rom_dir)), $(vc_gbc))
-	(cd $(basename $(basename $@))/; \
-	    makerom -f cxi -o ../../$(basename $@).cxi -rsf ../game.rsf \
-	            -exheader exheader.bin \
-	            -logo logo.lz \
-	            -plainrgn plain.bin \
-	            -code exefs/code.bin \
-	            -icon exefs/icon.bin \
-	            -banner exefs/banner.bin \
-	)
-
-$(vc_manual_cfa): $(vc_dir)/manual.rsf
-	makerom -f cfa -o $@ -rsf $(vc_dir)/manual.rsf
-
-$(vc_cia): $(vc_game_cxi) $(vc_manual_cfa)
-	makerom -f cia -o $@ -content $(basename $@).game.cxi:0:0 -content $(basename $@).manual.cfa:1:1
\ No newline at end of file
+	ctrtool --contents=$@contents $< >/dev/null
+	ctrtool --seeddb=${vc_dir}/seeddb.bin \
+	        --exheader=$@exheader.bin \
+	        --exefsdir=$@exefs \
+	        --romfsdir=$@romfs \
+	        --logo=$@logo.lz \
+	        --plainrgn=$@plain.bin \
+	        $@contents.0000.* >/dev/null
+	ctrtool --seeddb=${vc_dir}/seeddb.bin \
+	        --romfsdir=$@manual \
+	        $@contents.0001.* >/dev/null
+	rm -f $@contents.*
+	rm -f $@romfs/rom/*
+	rm -f $@romfs/*.patch
+
+# romfs files have the pattern appear twice in the path, which breaks pattern rules; we have to use `eval` instead
+# Careful that the contents of the `define`s are expanded twice:
+# 1. In the `call` function, and
+# 2. In the `eval` function.
+
+define copy_patch_rule
+$${vc_dir}/$(1)/romfs/$(1).patch: $(1).patch | ${vc_dir}/$(1)/
+	mkdir -p $${@D}
+	cp -T $$< $$@
+endef
+$(foreach vc,${vc_name},$(eval $(call copy_patch_rule,${vc})))
+
+define copy_rom_rule
+$${vc_dir}/$(1)/romfs/rom/$(1): $(1).gbc | ${vc_dir}/$(1)/
+	mkdir -p $${@D}
+	cp -T $$< $$@
+endef
+$(foreach vc,${vc_name},$(eval $(call copy_rom_rule,${vc})))
+
+define make_cxi_rule
+$${vc_dir}/$(1).game.cxi: $${vc_dir}/game.rsf $${vc_dir}/$(1)/romfs/$(1).patch $${vc_dir}/$(1)/romfs/rom/$(1)
+	makerom -f cxi -o $$@ -rsf $$< \
+	        -exheader $${vc_dir}/$(1)/exheader.bin \
+	        -logo $${vc_dir}/$(1)/logo.lz \
+	        -plainrgn $${vc_dir}/$(1)/plain.bin \
+	        -code $${vc_dir}/$(1)/exefs/code.bin \
+	        -icon $${vc_dir}/$(1)/exefs/icon.bin \
+	        -banner $${vc_dir}/$(1)/exefs/banner.bin
+endef
+$(foreach vc,${vc_name},$(eval $(call make_cxi_rule,${vc})))
+
+${vc_dir}/%.manual.cfa: ${vc_dir}/manual.rsf
+	makerom -f cfa -o $@ -rsf $<
+
+${vc_dir}/%.cia: ${vc_dir}/%.game.cxi ${vc_dir}/%.manual.cfa
+	makerom -f cia -o $@ -content $<:0:0 -content ${vc_dir}/$*.manual.cfa:1:1
```

Would this be better if it were like this:

```suggestion
define make_cia_rule
$${vc_dir}/$(1).cia: ${vc_dir}/$(1).game.cxi ${vc_dir}/$(1).manual.cfa
	makerom -f cia -o $$@ -content $$<:0:0 -content ${vc_dir}/$(1).manual.cfa:1:1
endef
$(foreach vc,${vc_name},$(eval $(call make_cia_rule,${vc})))
```

Just curious

---

Yes and no. Implicit rules are IMO easier to read, but they also generate less user-friendly errors when their prereqs are missing (you'll get "no rules for X" instead of "missing prereq Y for X", due to their nature).